nexusparser/tools/yaml2nxdl/yaml2nxdl_recursive_build.py

```diff
@@ -1,226 +1,246 @@
+def xml_handle_link(obj, keyword, value):
+    # if we have a link we decode the name attribute from <optional string>(link)[:-6]
+    if len(keyword[:-6]) >= 1 and isinstance(value, dict) and 'target' in value.keys():
+        if isinstance(value['target'], str) and len(value['target']) >= 1:
+            lnk = ET.SubElement(obj, 'link')
+            lnk.set('name', keyword[:-6])
+            lnk.set('target', value['target'])
         else:
-            raise ValueError(k+' value for target member of a link is invalid !')
+            raise ValueError(keyword + ' value for target member of a link is invalid !')
     else:
-        raise ValueError(k+' the formatting of what seems to be a link is invalid in the yml file !')
+        raise ValueError(keyword + ' the formatting of what seems to be a link is invalid in the yml file !')
 
 
 def recursive_build(obj, dct):
     """
-    obj is the current node where we want to append to, dct is a dictionary object which represents the content of the child in the nested set of dictionaries
+    obj is the current node where we want to append to,
+    dct is a dictionary object which represents the content of the child in the nested set of dictionaries
     """
-    for k, v in iter(dct.items()):
-        print('Processing key '+k+' value v is a dictionary '+str(isinstance(v,dict)))
-        #base class prefix tag removal
-        #k = '(NXentry)'
-        #k = 'sensor_size(NX_INT)'
-        kName, kType = nx_name_type_resolving(k)
-        print('kName: '+kName+' kType: '+kType)
+    for keyword, value in iter(dct.items()):
+        kName, kType = nx_name_type_resolving(keyword)
+        print('kName: ' + kName + ' kType: ' + kType)
+        print(type(kName),type(kType),type(keyword))
         if kName == '' and kType == '':
+            print('all empty')
             raise ValueError('ERROR: Found an improper YML key !')
-        elif kType.replace('NX','nx_') in nx_base_clss or kType.replace('NX','nx_') in nx_cand_clss:
-            #we can be sure we need to instantiate a new group
-            grp = etree.SubElement(obj, 'group')
-            if kName != '': #use the custom name for the group
-                grp.set('name', kName) #because we are a base or cand clss kName is a NX<str>
-            #else: #use the NeXus default to infer the name from the base/cand class name stripping the 'NX' decorator prefix
-            #    grp.set('name', kName[2:])
+        elif keyword[-6:] == '(link)': 
+            print('link elif')
+            xml_handle_link(obj, keyword, value)
+        elif (kType in nx_clss) or (kType not in nx_type_keys + ['']):
+            print('nx_class')
+            # we can be sure we need to instantiate a new group
+            grp = ET.SubElement(obj, 'group')
+            if kName != '':  # use the custom name for the group
+                grp.set('name', kName)
+            else:
+                if kType != '':
+                    grp.set('name', kType)
```

Comment:
I think this else block should be removed, and no name defined if the object is not named (see my other comment)